### ISSUE TITLE
LUKS key file stopped working after upgrading to 248

### ISSUE THREAD

**Issue by Frederick888** [systemd/systemd#19193]
**systemd version the issue has been seen with**
 > 248-1

**Used distribution**
 > Arch Linux

**Linux kernel version used** (`uname -a`)

 > `Linux 5.11.11-arch1-1 #1 SMP PREEMPT Tue, 30 Mar 2021 14:10:17 +0000 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > x86-64

**Expected behaviour you didn't see**
 > LUKS devices listed in `/etc/crypttab` with key files are opened normally as usual

**Unexpected behaviour you saw**
Devices in `/etc/crypttab` are not opened automatically. Manually running `cryptsetup open --type=luks --key-file=/etc/crypttab.key /dev/disk/by-uuid/xxx Name` works all fine.

**Steps to reproduce the problem**
```
# /etc/crypttab
crypthdd500     UUID=yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy       /etc/crypttab.key       luks,noauto
crypthdd2000    UUID=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx       /etc/crypttab.key       luks,noauto
```

**Additional program output to the terminal or log subsystem illustrating the issue**

```
Apr 03 22:31:15 systemd[1]: Starting Cryptography Setup for crypthdd2000...
Apr 03 22:31:15 systemd[1]: Found device Hitachi_HTXXXXXXXXXXXXX.
Apr 03 22:31:15 systemd[1]: Starting Cryptography Setup for crypthdd500...
Apr 03 22:31:15 systemd-cryptsetup[456]: Set cipher aes, mode xts-plain64, key size 512 bits for device /dev/disk/by-uuid/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.
Apr 03 22:31:15 systemd-cryptsetup[456]: Failed to activate with key file '/etc/crypttab.key': Invalid argument
Apr 03 22:31:15 systemd-cryptsetup[455]: Set cipher aes, mode xts-plain64, key size 512 bits for device /dev/disk/by-uuid/yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy.
Apr 03 22:31:15 systemd-cryptsetup[455]: Failed to activate with key file '/etc/crypttab.key': Invalid argument
Apr 03 22:31:15 systemd[1]: systemd-cryptsetup@crypthdd2000.service: Main process exited, code=exited, status=1/FAILURE
Apr 03 22:31:15 systemd[1]: systemd-cryptsetup@crypthdd2000.service: Failed with result 'exit-code'.
Apr 03 22:31:15 systemd[1]: Failed to start Cryptography Setup for crypthdd2000.
Apr 03 22:31:15 systemd[1]: Dependency failed for /dev/mapper/crypthdd2000.
Apr 03 22:31:15 systemd[1]: systemd-cryptsetup@crypthdd500.service: Main process exited, code=exited, status=1/FAILURE
Apr 03 22:31:15 systemd[1]: systemd-cryptsetup@crypthdd500.service: Failed with result 'exit-code'.
Apr 03 22:31:15 systemd[1]: Failed to start Cryptography Setup for crypthdd500.
```

After downgrading to `247.4-2` the issue is gone.

Relevant Arch ticket (not by me): https://bugs.archlinux.org/task/70272

**Comment by Sinbiosis** [comment]
I'm commenting to confirm that I also have the same issues understand the same configurations and system software versions. Downgrading to systemd 247.3-1 (which was my last installed version) fixed the issue.

**Comment by Slenderchat** [comment]
I'm also commenting to confirm that I also have the same issues under the same configurations and system software versions. Downgrading to systemd 247.4-2 (which was my last installed version) fixed the issue.

**Comment by mmhat** [comment]
I can confirm this issue but my error message is different:
```
Failed to load key file 'disk.key': Argument list too long
```
Downgrading to 247 helps. Manually opening the device with cryptsetup works fine.

**Comment by kelppo** [comment]
Commit https://github.com/systemd/systemd/commit/b8c80b56d13ba2704ae28bcf1ada4d3f512cb1be added an if branch checking pass_volume_key to this piece of code that wasn't there previously:
https://github.com/systemd/systemd/blob/b997d1115bf21b58dd876e02ce095bc483e3c875/src/cryptsetup/cryptsetup.c#L810

I can't make sure which branch is taken since I don't unlock my LUKS disk with a key file so can anyone test this patch? It has at least been compile tested:

```
diff --git a/src/cryptsetup/cryptsetup.c b/src/cryptsetup/cryptsetup.c
index ee31400e3f..83ef741078 100644
--- a/src/cryptsetup/cryptsetup.c
+++ b/src/cryptsetup/cryptsetup.c
@@ -1181,8 +1181,7 @@ static int attach_luks_or_plain_or_bitlk_by_key_file(
                 struct crypt_device *cd,
                 const char *name,
                 const char *key_file,
-                uint32_t flags,
-                bool pass_volume_key) {
+                uint32_t flags) {
 
         _cleanup_(erase_and_freep) char *kfdata = NULL;
         _cleanup_free_ char *bindname = NULL;
@@ -1210,10 +1209,7 @@ static int attach_luks_or_plain_or_bitlk_by_key_file(
                 return -EAGAIN; /* Log actual error, but return EAGAIN */
         }
 
-        if (pass_volume_key)
-                r = crypt_activate_by_volume_key(cd, name, kfdata, kfsize, flags);
-        else
-                r = crypt_activate_by_passphrase(cd, name, arg_key_slot, kfdata, kfsize, flags);
+        r = crypt_activate_by_passphrase(cd, name, arg_key_slot, kfdata, kfsize, flags);
         if (r == -EPERM) {
                 log_error_errno(r, "Failed to activate with key file '%s'. (Key data incorrect?)", key_file);
                 return -EAGAIN; /* Log actual error, but return EAGAIN */
@@ -1334,7 +1330,7 @@ static int attach_luks_or_plain_or_bitlk(
         if (key_data)
                 return attach_luks_or_plain_or_bitlk_by_key_data(cd, name, key_data, key_data_size, flags, pass_volume_key);
         if (key_file)
-                return attach_luks_or_plain_or_bitlk_by_key_file(cd, name, key_file, flags, pass_volume_key);
+                return attach_luks_or_plain_or_bitlk_by_key_file(cd, name, key_file, flags);
 
         return attach_luks_or_plain_or_bitlk_by_passphrase(cd, name, passwords, flags, pass_volume_key);
 }
```

**Comment by poettering** [comment]
Hmm, anyone would be willing to bisect this?

**Comment by Frederick888** [comment]
@kelppo No it didn't work for me unfortunately.

@poettering I tried but it seems that the latest Meson configuration in Arch didn't work for older versions (or maybe just a compilation error?)... haven't got time atm to dig deeper. FWIW I noticed that downgrading systemd after boot works already (no need to rebuild initramfs, reboot, etc). Could you please suggest which binary one should replace to bisect this? It'll be much quicker than rebuilding the Arch package.

**Comment by aaronmaslen** [comment]
@Frederick888 I'd guess systemd-cryptsetup is the binary you want.

**Comment by fedevx** [comment]
This also affects environments were multiple Luks volumes are encrypted with the same key. Systemd seems to be storing the passphrase temporarily in a file. As a result, the first drive unlocks successfully with the provided passphrase (likely reading from memory) but subsequent drives fail to unlock (likely when trying to read from the file). The logs show the below error:

```
May 11 07:41:21 archlinux systemd-tty-ask-password-agent[299]: Failed to send: No such file or directory
May 11 07:41:21 archlinux systemd-tty-ask-password-agent[299]: Invalid password file /run/systemd/ask-password/ask.kLSJTl
May 11 07:41:21 archlinux systemd-tty-ask-password-agent[299]: Failed to process password: No such file or directory
```

**Comment by poettering** [comment]
So I tried hard to reproduce the issue, but no luck. I have the suspicion there are different issues at hand here, i.e. one EINVAL thing, and another E2BIG thing, so maybe we shouldn't mix up these two issues.

I did fine one bug about a missing error path, which is fixed by #19772. 

Would be great if anyone could try to reproduce this with current git main, with #19772 applied, and then would try to run the cryptsetup command from the command line, like this:

```
SYSTEMD_LOG_LEVEL=debug /usr/lib/systemd/systemd-cryptsetup attach crypthdd500 UUID=yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy       /etc/crypttab.key       luks,noauto
```

the final four args are 1:1 the entries from crypttab.

(of course when running this make sure to have the volume detatched)

**Comment by poettering** [comment]
Dropping from milestone, since there's no feedback so far, hence maybe the issue got fixed already.

Please report back if the changes in git main made it work, and if not, please provide me with the requested log, or let me know a good, minimal way I can reproduce the issue

**Comment by eworm-de** [comment]
As this has been reported for Arch Linux... Anybody interested in packages from current `HEAD`?

**Comment by Slenderchat** [comment]
Hello, I am the one who reported this bug initially on Arch Linux bugtracker. I have plenty of setups where i can test whether this bug still appears. I will test and provide feedback within next few days.

**Comment by Frederick888** [comment]
Sorry, been procrastinating too much recently...

First of all, I built and installed 00fb6caf70c4ac742a5b4e913494e7d431abf0ec using https://aur.archlinux.org/packages/systemd-git/ but it didn't fix my issue.

Then I bisected it:

```
# bad: [938bdfc0fa737d86eb3ecc70506e11e5f740e0dc] Merge pull request #19157 from keszybz/read-medium-sized-virtual-file
# good: [4d484e14bb9864cef1d124885e625f33bf31e91c] Merge pull request #17738 from keszybz/hwdb-news-update-v247
git bisect start 'v248' 'v247'
# good: [edf370af9e9fafad01393699e7a6f34bf0568dd6] Merge pull request #18432 from yuwata/libude-list-cleanups
git bisect good edf370af9e9fafad01393699e7a6f34bf0568dd6
# good: [b332778b30d23193c792d5f5c5dcccd61f4a489c] Merge pull request #18684 from yuwata/sd-device-cleanups-and-udevadm-fix
git bisect good b332778b30d23193c792d5f5c5dcccd61f4a489c
# good: [2541462f1b6a4a3985c3b2db181c2ab61a9efefb] Merge pull request #18890 from keszybz/fuzz-bus-match
git bisect good 2541462f1b6a4a3985c3b2db181c2ab61a9efefb
# good: [b1eea703e01da1e280e179fb119449436a0c9b8e] resolved: don't flush answer RRs on CNAME redirect too early
git bisect good b1eea703e01da1e280e179fb119449436a0c9b8e
# good: [dfb433f2a77a04d44aaa64d8364805e36cb25f5a] firewall-util: probe firewall backend in fw_ctx_new()
git bisect good dfb433f2a77a04d44aaa64d8364805e36cb25f5a
# good: [3b0754b16c5535fd48f48ca39f69335ef695efe7] config files: recommend systemd-analyze cat-config
git bisect good 3b0754b16c5535fd48f48ca39f69335ef695efe7
# bad: [063a43a1a7e29a6db97c5282435f716d06274fb4] Revert "fix wrong statement JOURNAL_FILE_FORMAT.md doc"
git bisect bad 063a43a1a7e29a6db97c5282435f716d06274fb4
# good: [a81c7ac8d408a2618d488e708b40530bcdad6bd1] Merge pull request #19129 from keszybz/test-random-range
git bisect good a81c7ac8d408a2618d488e708b40530bcdad6bd1
# bad: [bc52deda4b0e354949c0783afd5e9c6f2034b8b2] tests: drop calls to unsetenv SYSTEMD_MEMPOOL
git bisect bad bc52deda4b0e354949c0783afd5e9c6f2034b8b2
# bad: [a9899ff3588801b825322d40564b271c7216fe46] basic/fileio: optimize buffer sizes in read_full_virtual_file()
git bisect bad a9899ff3588801b825322d40564b271c7216fe46
# good: [ca79564309e87f1cd92f7c79a03fccb792cca274] basic/fileio: simplify calculation of buffer size in read_full_virtual_file()
git bisect good ca79564309e87f1cd92f7c79a03fccb792cca274
# first bad commit: [a9899ff3588801b825322d40564b271c7216fe46] basic/fileio: optimize buffer sizes in read_full_virtual_file()
```


`PKGBUILD` I used for bisecting based on systemd-git@b415dbb33f0a98bd40aab3df8dc6f6805dc28c04 in case anyone's interested:

```PKGBUILD
# Maintainer: Yurii Kolesnykov <root@yurikoles.com>
# Based on core/systemd by:
# Maintainer: Christian Hesse <mail@eworm.de>
# Maintainer: Dave Reisner <dreisner@archlinux.org>
# Maintainer: Tom Gundersen <teg@jklm.no>

_pkgbase=systemd
pkgbase=$_pkgbase-git
pkgname=('systemd-git' 'systemd-libs-git' 'systemd-resolvconf-git' 'systemd-sysvcompat-git')
pkgdesc='systemd (git version)'
pkgver=248.rc4.r56.gca79564309
pkgrel=1
arch=('x86_64')
url='https://www.github.com/systemd/systemd'
makedepends=('acl' 'cryptsetup' 'docbook-xsl' 'gperf' 'lz4' 'xz' 'pam' 'libelf'
             'intltool' 'iptables' 'kmod' 'libcap' 'libidn2' 'libgcrypt'
             'libmicrohttpd' 'libxcrypt' 'libxslt' 'util-linux' 'linux-api-headers'
             'python-lxml' 'quota-tools' 'shadow' 'gnu-efi-libs' 'git'
             'meson' 'libseccomp' 'pcre2' 'audit' 'kexec-tools' 'libxkbcommon'
             'bash-completion' 'p11-kit' 'systemd')
options=('strip')
source=('git+https://github.com/systemd/systemd'
        '0001-Use-Arch-Linux-device-access-groups.patch'
        'initcpio-hook-udev'
        'initcpio-install-systemd'
        'initcpio-install-udev'
        'arch.conf'
        'loader.conf'
        'splash-arch.bmp'::'https://git.archlinux.org/svntogit/packages.git/plain/trunk/splash-arch.bmp?h=packages/systemd'
        'systemd-user.pam'
        'systemd-hook'
        '20-systemd-sysusers.hook'
        '30-systemd-binfmt.hook'
        '30-systemd-catalog.hook'
        '30-systemd-daemon-reload.hook'
        '30-systemd-hwdb.hook'
        '30-systemd-sysctl.hook'
        '30-systemd-tmpfiles.hook'
        '30-systemd-udev-reload.hook'
        '30-systemd-update.hook')
sha512sums=('SKIP'
            '73ec374130597eb9cfa9defa62574a6815428cf86a51ff9c7ce76a20d8d762c5258c066b15f0327a7bfcbb9c49fa7da6bae58be6ae838c64d28ca24edadeff65'
            'f0d933e8c6064ed830dec54049b0a01e27be87203208f6ae982f10fb4eddc7258cb2919d594cbfb9a33e74c3510cfd682f3416ba8e804387ab87d1a217eb4b73'
            '8e76f8334b95ce7fee9190f4a1016b16109f3a75b68635fc227b2b4791cf8179ef09b532b66b4ed885ddf98ed76befed3106f3c3088f1819ed8cdf4c13e0805a'
            'a25b28af2e8c516c3a2eec4e64b8c7f70c21f974af4a955a4a9d45fd3e3ff0d2a98b4419fe425d47152d5acae77d64e69d8d014a7209524b75a81b0edb10bf3a'
            '61032d29241b74a0f28446f8cf1be0e8ec46d0847a61dadb2a4f096e8686d5f57fe5c72bcf386003f6520bc4b5856c32d63bf3efe7eb0bc0deefc9f68159e648'
            'c416e2121df83067376bcaacb58c05b01990f4614ad9de657d74b6da3efa441af251d13bf21e3f0f71ddcb4c9ea658b81da3d915667dc5c309c87ec32a1cb5a5'
            '5a1d78b5170da5abe3d18fdf9f2c3a4d78f15ba7d1ee9ec2708c4c9c2e28973469bc19386f70b3cf32ffafbe4fcc4303e5ebbd6d5187a1df3314ae0965b25e75'
            'b90c99d768dc2a4f020ba854edf45ccf1b86a09d2f66e475de21fe589ff7e32c33ef4aa0876d7f1864491488fd7edb2682fc0d68e83a6d4890a0778dc2d6fe19'
            '869dab2b1837c964add4019bb402e24e52dbb7f009850ca69fcc5deddd923eeb98eb8ee38601f6e31531f30322472fe7df09af84df27f0467708406c55885323'
            '299dcc7094ce53474521356647bdd2fb069731c08d14a872a425412fcd72da840727a23664b12d95465bf313e8e8297da31259508d1c62cc2dcea596160e21c5'
            '0d6bc3d928cfafe4e4e0bc04dbb95c5d2b078573e4f9e0576e7f53a8fab08a7077202f575d74a3960248c4904b5f7f0661bf17dbe163c524ab51dd30e3cb80f7'
            '2b50b25e8680878f7974fa9d519df7e141ca11c4bfe84a92a5d01bb193f034b1726ea05b3c0030bad1fbda8dbb78bf1dc7b73859053581b55ba813c39b27d9dc'
            '63e55b3acd14bc54320b6f2310b43398651ad4e262d4f4a0135e05d34a993e56ed673cc46e57f15b418371df5c4cef6f54486db96325e4abb1d33fb1a3946254'
            'a1661ab946c6cd7d3c6251a2a9fd68afe231db58ce33c92c42594aedb5629be8f299ba08a34713327b373a3badd1554a150343d8d3e5dfb102999c281bd49154'
            '9426829605bbb9e65002437e02ed54e35c20fdf94706770a3dc1049da634147906d6b98bf7f5e7516c84068396a12c6feaf72f92b51bdf19715e0f64620319de'
            'da7a97d5d3701c70dd5388b0440da39006ee4991ce174777931fea2aa8c90846a622b2b911f02ae4d5fffb92680d9a7e211c308f0f99c04896278e2ee0d9a4dc'
            'a50d202a9c2e91a4450b45c227b295e1840cc99a5e545715d69c8af789ea3dd95a03a30f050d52855cabdc9183d4688c1b534eaa755ebe93616f9d192a855ee3'
            '825b9dd0167c072ba62cabe0677e7cd20f2b4b850328022540f122689d8b25315005fa98ce867cf6e7460b2b26df16b88bb3b5c9ebf721746dce4e2271af7b97')
prepare() {
  cd "$_pkgbase"

  # although removing _build folder in build() function feels more natural,
  # that interferes with the spirit of makepkg --noextract
  if [ -d _build ]; then
    rm -rf _build
  fi

  # Replace cdrom/dialout/tape groups with optical/uucp/storage
  #patch -Np1 -i ../0001-Use-Arch-Linux-device-access-groups.patch
}

pkgver() {
  cd "$_pkgbase"
  git describe --long --tags | sed 's/^v//;s/\([^-]*-g\)/r\1/;s/-/./g'
}

build() {
  local _timeservers=({0..3}.arch.pool.ntp.org)
  local _nameservers=(
    # We use these public name services, ordered by their
    # privacy policy (hopefully):
    #  * Cloudflare (https://1.1.1.1/)
    #  * Quad9 without filtering (https://www.quad9.net/)
    #  * Google (https://developers.google.com/speed/public-dns/)
    1.1.1.1
    9.9.9.10
    8.8.8.8
    2606:4700:4700::1111
    2620:fe::10
    2001:4860:4860::8888
  )

  local _meson_options=(
    -Dversion-tag="${pkgver}-${pkgrel}-arch"
    -Dmode=release

    -Dgnu-efi=true
    -Dima=false
    -Dlibidn2=true
    -Dlz4=true
    -Dman=true

    # We disable DNSSEC by default, it still causes trouble:
    # https://github.com/systemd/systemd/issues/10579
    
    -Ddbuspolicydir=/usr/share/dbus-1/system.d
    -Ddefault-dnssec=no
    -Ddefault-hierarchy=hybrid
    -Ddefault-kill-user-processes=false
    -Ddefault-locale=C
    -Ddns-over-tls=openssl
    -Dfallback-hostname='archlinux'
    -Dnologin-path=/usr/bin/nologin
    -Dntp-servers="${_timeservers[*]}"
    -Ddns-servers="${_nameservers[*]}"
    -Drpmmacrosdir=no
    -Dsysvinit-path=
    -Dsysvrcnd-path=
  )

  # arch-meson "$_pkgbase" build "${_meson_options[@]}"

  git -C systemd cp --no-commit 95ba433a5f34baf92921fb58051bc8241f908c0e
  git -C systemd merge-base --is-ancestor a71c09685021cbcecb7566a00342421f635cc002 HEAD && git -C systemd cp --no-commit ab1aa6368a883bce88e3162fee2bea14aacedf23
  ninja -C build
}

check() {
  meson test -C build
}

package_systemd-git() {
  pkgdesc='system and service manager (git version)'
  license=('GPL2' 'LGPL2.1')
  depends=('acl' 'libacl.so' 'bash' 'cryptsetup' 'libcryptsetup.so' 'dbus'
           'iptables' 'kbd' 'kmod' 'libkmod.so' 'hwids' 'libcap' 'libcap.so'
           'libgcrypt' 'libxcrypt' 'libcrypt.so' 'systemd-libs-git' 'libidn2' 'lz4' 'pam'
           'libelf' 'libseccomp' 'libseccomp.so' 'util-linux' 'libblkid.so'
           'libmount.so' 'xz' 'pcre2' 'audit' 'libaudit.so' 'libp11-kit'
           'libp11-kit.so' 'openssl')
  provides=("systemd=$pkgver" 'nss-myhostname' "systemd-tools=$pkgver" "udev=$pkgver")
  replaces=('nss-myhostname' 'systemd-tools' 'udev')
  conflicts=('systemd' 'nss-myhostname' 'systemd-tools' 'udev')
  optdepends=('libmicrohttpd: remote journald capabilities'
              'quota-tools: kernel-level quota management'
              'systemd-sysvcompat: symlink package to provide sysvinit binaries'
              'polkit: allow administration as unprivileged user'
              'curl: machinectl pull-tar and pull-raw')
  backup=(etc/pam.d/systemd-user
          etc/systemd/coredump.conf
          etc/systemd/homed.conf
          etc/systemd/journald.conf
          etc/systemd/journal-remote.conf
          etc/systemd/journal-upload.conf
          etc/systemd/logind.conf
          etc/systemd/networkd.conf
          etc/systemd/pstore.conf
          etc/systemd/resolved.conf
          etc/systemd/sleep.conf
          etc/systemd/system.conf
          etc/systemd/timesyncd.conf
          etc/systemd/user.conf
          etc/udev/udev.conf)
  install=systemd.install

  DESTDIR="$pkgdir" meson install -C build

  # we'll create this on installation
  rmdir "$pkgdir"/var/log/journal/remote

  # runtime libraries shipped with systemd-libs
  install -d -m0755 systemd-libs
  mv "$pkgdir"/usr/lib/lib{nss,systemd,udev}*.so* systemd-libs

  # manpages shipped with systemd-sysvcompat
  rm "$pkgdir"/usr/share/man/man8/{halt,poweroff,reboot,shutdown}.8

  # executable (symlinks) shipped with systemd-sysvcompat
  rm "$pkgdir"/usr/bin/{halt,init,poweroff,reboot,shutdown}

  # files shipped with systemd-resolvconf
  rm "$pkgdir"/usr/{bin/resolvconf,share/man/man1/resolvconf.1}

  # avoid a potential conflict with [core]/filesystem
  rm "$pkgdir"/usr/share/factory/etc/{issue,nsswitch.conf}
  sed -i -e '/^C \/etc\/nsswitch\.conf/d' \
    -e '/^C \/etc\/issue/d' "$pkgdir"/usr/lib/tmpfiles.d/etc.conf

  # add back tmpfiles.d/legacy.conf, normally omitted without sysv-compat
  install -m0644 "$_pkgbase"/tmpfiles.d/legacy.conf "$pkgdir"/usr/lib/tmpfiles.d

  # ship default policy to leave services disabled
  echo 'disable *' >"$pkgdir"/usr/lib/systemd/system-preset/99-default.preset

  # add mkinitcpio hooks
  install -D -m0644 initcpio-install-systemd "$pkgdir"/usr/lib/initcpio/install/systemd
  install -D -m0644 initcpio-install-udev "$pkgdir"/usr/lib/initcpio/install/udev
  install -D -m0644 initcpio-hook-udev "$pkgdir"/usr/lib/initcpio/hooks/udev

  # The group 'systemd-journal' is allocated dynamically and may have varying
  # gid on different systems. Let's install with gid 0 (root), systemd-tmpfiles
  # will fix the permissions for us. (see /usr/lib/tmpfiles.d/systemd.conf)
  install -d -o root -g root -m 2755 "$pkgdir"/var/log/journal

  # match directory owner/group and mode from [extra]/polkit
  install -d -o root -g 102 -m 0750 "$pkgdir"/usr/share/polkit-1/rules.d

  # add example bootctl configuration
  install -D -m0644 arch.conf "$pkgdir"/usr/share/systemd/bootctl/arch.conf
  install -D -m0644 loader.conf "$pkgdir"/usr/share/systemd/bootctl/loader.conf
  install -D -m0644 splash-arch.bmp "$pkgdir"/usr/share/systemd/bootctl/splash-arch.bmp

  # pacman hooks
  install -D -m0755 systemd-hook "$pkgdir"/usr/share/libalpm/scripts/systemd-hook
  install -D -m0644 -t "$pkgdir"/usr/share/libalpm/hooks *.hook

  # overwrite the systemd-user PAM configuration with our own
  install -D -m0644 systemd-user.pam "$pkgdir"/etc/pam.d/systemd-user
}

package_systemd-libs-git() {
  pkgdesc='systemd client libraries (git version)'
  depends=('glibc' 'libcap' 'libgcrypt' 'lz4' 'xz' 'zstd')
  license=('LGPL2.1')
  provides=("systemd-libs=$pkgver" 'libsystemd' 'libsystemd.so' 'libudev.so')
  conflicts=('systemd-libs' 'libsystemd')
  replaces=('libsystemd')

  install -d -m0755 "$pkgdir"/usr
  mv systemd-libs "$pkgdir"/usr/lib
}

package_systemd-resolvconf-git() {
  pkgdesc='systemd resolvconf replacement (for use with systemd-resolved, git version)'
  license=('LGPL2.1')
  depends=('systemd-git')
  provides=("systemd-resolvconf=$pkgver" 'openresolv' 'resolvconf')
  conflicts=('systemd-resolvconf' 'openresolv')

  install -d -m0755 "$pkgdir"/usr/bin
  ln -s resolvectl "$pkgdir"/usr/bin/resolvconf

  install -d -m0755 "$pkgdir"/usr/share/man/man1
  ln -s resolvectl.1.gz "$pkgdir"/usr/share/man/man1/resolvconf.1.gz
}

package_systemd-sysvcompat-git() {
  pkgdesc='sysvinit compat for systemd (git version)'
  license=('GPL2')
  conflicts=('systemd-sysvcompat' 'sysvinit')
  depends=('systemd-git')
  provides=("systemd-sysvcompat=$pkgver")

  install -D -m0644 -t "$pkgdir"/usr/share/man/man8 \
    build/man/{halt,poweroff,reboot,shutdown}.8

  install -d -m0755 "$pkgdir"/usr/bin
  ln -s ../lib/systemd/systemd "$pkgdir"/usr/bin/init
  for tool in halt poweroff reboot shutdown; do
    ln -s systemctl "$pkgdir"/usr/bin/$tool
  done
}
```

**Comment by Frederick888** [comment]
Found the issue (I think).

I had been using a key file which was exactly 4MB, which was the old read limit `4U*1024U*1024U`.

In a9899ff3588801b825322d40564b271c7216fe46, the limit was updated to `4U*1024U*1024U - 1`, so now https://github.com/systemd/systemd/blob/a9899ff3588801b825322d40564b271c7216fe46/src/basic/fileio.c#L411-L412 returns E2BIG.

I just switched to a smaller key file and upgraded systemd to 248.3-2, and now all my disks can be unlocked normally.

FYI the default key file size limit in cryptsetup is 8192KB: https://gitlab.com/cryptsetup/cryptsetup/blob/cd3cb945ab24ef819da30c90877b0cf9dd90c2e5/configure.ac#L623.
